File: codingtest/java lv1~/가장_많이_받은_선물.java
import java.util.*;
class Solution {
    int GIVE = 0;
    int TAKE = 1;

    public int solution(String[] friends, String[] gifts) {
        Map<String, Integer> idxMap = new HashMap<>();
        int idx = 0;

        for(String friend : friends) {
            idxMap.put(friend, idx++);
        }

        int[][] matrix = new int[friends.length][friends.length];
        int[][] giveAndTakeArr = new int[friends.length][2];
        String[] names;
        for(String gift : gifts) {
            names = gift.split(" ");

            int A = idxMap.get(names[0]);
            int B = idxMap.get(names[1]);

            matrix[A][B]++;
            giveAndTakeArr[A][GIVE]++;
            giveAndTakeArr[B][TAKE]++;
        }

        Map<Integer, Integer> answerMap = new HashMap<>();
        for(int i=0; i<matrix.length-1; i++) {
            for(int j=i+1; j<matrix[i].length; j++) {

                // A, B 간에 준 선물 숫자
                int A = matrix[i][j];
                int B = matrix[j][i];

                if((A == 0 && B == 0) || (A == B)) {
                    int AVal = giveAndTakeArr[i][GIVE] - giveAndTakeArr[i][TAKE];
                    int BVal = giveAndTakeArr[j][GIVE] - giveAndTakeArr[j][TAKE];

                    if(AVal == BVal) {
                        continue;
                    }

                    if(AVal > BVal) {
                        answerMap.put(i, answerMap.getOrDefault(i, 0) +1);
                    }else {
                        answerMap.put(j, answerMap.getOrDefault(j, 0) +1);
                    }
                }else {
                    if(A > B) {
                        answerMap.put(i, answerMap.getOrDefault(i, 0) +1);
                    }else {
                        answerMap.put(j, answerMap.getOrDefault(j, 0) +1);
                    }
                }
            }
        }

        int answer = 0;
        for(Map.Entry<Integer, Integer> entry : answerMap.entrySet()) {
            answer = Math.max(answer, entry.getValue());
        }

        return answer;
    }
}

File: codingtest/pyrhon lv1~/가장 많이 받은 선물.py
def solution(friends, gifts):
    friends.sort()
    result = {}
    bonus = {}
    gift_point = {}
    next_month = {}

    for i in friends:
        gift_point[i] = 0
        next_month[i] = 0
        result[i] = 0
        a = friends.copy()
        a.remove(i)
        for aa in a:
            if i < aa:
                bonus[i+' '+aa] = [0, 0]
            else:
                bonus[aa + ' ' + i] = [0, 0]
        # 선물지수 dict

    for i in gifts:
        a = i.split(" ")
        result[a[0]] += 1
        # 선물지수
        gift_point[a[0]] += 1
        gift_point[a[1]] -= 1
        if a[0] < a[1]:
            bonus[a[0] + ' ' + a[1]][0] += 1
        else:
            bonus[a[1] + ' ' + a[0]][1] += 1

    for i in bonus:
        a = i.split(' ')[0]
        b = i.split(' ')[1]
        if bonus[i][0] > bonus[i][1]:
            next_month[a] += 1
        if bonus[i][0] < bonus[i][1]:
            next_month[b] += 1
        if bonus[i][0] == bonus[i][1]:
            if gift_point[a] > gift_point[b]:
                next_month[a] += 1
            if gift_point[a] < gift_point[b]:
                next_month[b] += 1
    answer = 0

    for tt in next_month.values():
        if answer < tt:
            answer = tt

    return answer

